src/api.ts

```diff
@@ -100,33 +100,76 @@ export async function deleteImage(hash: CommitHash) {
 	}
 }
 
+const startingContainers: Map<CommitHash, Promise<ContainerInfo> > = new Map();
 export async function startContainer(commitHash: CommitHash) {
 	l.log({ commitHash }, `Starting up container for ${commitHash}`);
 	const image = getImageName(commitHash);
+
+	// do we have an existing container?
+	const existingContainer = getRunningContainerForHash( commitHash );
+	if ( existingContainer ) {
+		l.log( { commitHash, containerId: existingContainer.Id }, `Found a running container for ${commitHash}`);
+		return Promise.resolve( existingContainer );
+	}
+
+	// are we starting one already?
+	if( startingContainers.has( commitHash ) ) {
+		l.log( { commitHash, containerId: existingContainer.Id }, `Already starting a container for ${commitHash}`)
+		return startingContainers.get( commitHash );
+	}
+
+	const startPromise = _startContainer( image, commitHash );
+	startingContainers.set( commitHash, startPromise );
+	startPromise.then(
+		s => {
+			startingContainers.delete( commitHash );
+			return s;
+		},
+		err => {
+			startingContainers.delete( commitHash );
+			throw err;
+		}
+	)
+	return startPromise;
+}
+
+async function _startContainer( image: string, commitHash: CommitHash ): Promise<ContainerInfo> {
+	// ok, try to start one
 	let freePort: number;
 	try {
 		freePort = await portfinder.getPortPromise();
-	} catch (err) {
-		l.error({ err, commitHash }, `Error while attempting to find a free port for ${commitHash}`);
-		return;
+	} catch( err ) {
+		l.error( { err, image }, `Error while attempting to find a free port for ${image}`);
+		throw err;
 	}
 
 	const exposedPort = `${config.build.exposedPort}/tcp`;
-	docker.run(
-		image,
-		[],
-		process.stdout,
-		{
-			...config.build.containerCreateOptions,
-			ExposedPorts: { [exposedPort]: {} },
-			PortBindings: { [exposedPort]: [{ HostPort: freePort.toString() }] },
-			Tty: false,
+	const dockerPromise = new Promise( ( resolve, reject ) => {
+		docker.run(
+			image,
+			[],
+			process.stdout,
+			{
+				...config.build.containerCreateOptions,
+				ExposedPorts: { [exposedPort]: {} },
+				PortBindings: { [exposedPort]: [{ HostPort: freePort.toString() }] },
+				Tty: false,
+			},
+			(err, succ) => err
+				? reject( { error: err, freePort } )
+				: resolve( { success: succ, freePort } )
+		);
+	} )
+	return dockerPromise.then(
+		( { success, freePort } ) => {
+			l.log({ image, freePort }, `Successfully started container for ${image} on ${freePort}`);
+			return refreshRunningContainers().then( () => getRunningContainerForHash( commitHash ) );
```

Comment:
prior, this async process would "complete" and we'd rely on the timer to refresh the running containers. With the new lock in place, we can refresh the running containers before letting go of the lock, which makes sure state.containers is consistent.